scarb/src/ops/package.rs

```diff
@@ -88,10 +96,53 @@ fn before_package(ws: &Workspace<'_>) -> Result<()> {
     Ok(())
 }
 
-#[tracing::instrument(level = "trace", skip(_opts, ws))]
+#[derive(Serialize)]
+struct VcsInfo {
+    sha1: String,
+    path_in_vcs: String,
+}
+
+#[allow(clippy::dbg_macro)]
+fn extract_vcs_info(pkg: &Package, opts: &PackageOpts) -> Result<Option<VcsInfo>> {
+    if let Ok(repo) = gix::discover(pkg.root()) {
+        // There is nothing to parse if repository is empty, so HEAD reference needs to exist.
+        if let Ok(sha1) = repo.rev_parse_single("HEAD") {
+            // `git status -s` output is empty only if there are no changes at all, but always returns success.
+            // `git diff-index --quiet HEAD` returns status code correctly, but doesn't take untracked files into account.
+            let output = Command::new("git")
+                .current_dir(pkg.root())
+                .arg("status")
+                .arg("-s")
+                .output()?;
+
+            ensure!(
+                opts.allow_dirty || output.stdout.is_empty(),
+                indoc!(
+                    r#"
+                        cannot package a repository containing uncommited changes
+                        help: `--allow-dirty` to ignore this check
+                    "#
+                )
```

Comment:
```suggestion
                indoc! {r#"
                    cannot package a repository containing uncommited changes
                    help: to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag
                "#}
```

we need to emphasize that doing `--allow-dirty` is probably a bad idea!